File: backend/models/Product.js
const mongoose = require("mongoose");

const variantSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true
    },
    value: {
      type: String,
      required: true
    },
    price: {
      type: Number,
      required: true
    }
  },
  { _id: true }
);

const emiPlanSchema = new mongoose.Schema(
  {
    tenureMonths: {
      type: Number,
      required: true
    },
    monthlyAmount: {
      type: Number,
      required: true
    },
    totalAmount: {
      type: Number,
      required: true
    },
    interest: {
      type: Number,
      default: 0
    },
    isNoCost: {
      type: Boolean,
      default: true
    }
  },
  { _id: true }
);

const productSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true
    },

    brand: {
      type: String,
      required: true
    },

    category: {
      type: String,
      required: true
    },

    description: {
      type: String,
      required: true
    },

    imageUrl: {
      type: String,
      required: true
    },

    price: {
      type: Number,
      required: true
    },

    originalPrice: {
      type: Number
    },

    rating: {
      type: Number,
      default: 4.5
    },

    variants: [variantSchema],

    emiPlans: [emiPlanSchema],

    available: {
      type: Boolean,
      default: true
    }
  },
  {
    timestamps: true
  }
);

module.exports = mongoose.model("Product", productSchema);

File: lib/models/product.dart
import 'emi_plan.dart';
import 'product_variant.dart';

class Product {
  final String id;
  final String name;
  final String brand;
  final String category;
  final String description;
  final String imageUrl;
  final double price;
  final double? originalPrice;
  final double rating;
  final List<ProductVariant> variants;
  final List<EmiPlan> emiPlans;

  Product({
    required this.id,
    required this.name,
    required this.brand,
    required this.category,
    required this.description,
    required this.imageUrl,
    required this.price,
    this.originalPrice,
    required this.rating,
    required this.variants,
    required this.emiPlans,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'],
      name: json['name'],
      brand: json['brand'],
      category: json['category'],
      description: json['description'],
      imageUrl: json['imageUrl'],
      price: (json['price'] as num).toDouble(),
      originalPrice: json['originalPrice'] != null
          ? (json['originalPrice'] as num).toDouble()
          : null,
      rating: (json['rating'] as num).toDouble(),
      variants: (json['variants'] as List)
          .map((e) => ProductVariant.fromJson(e))
          .toList(),
      emiPlans: (json['emiPlans'] as List)
          .map((e) => EmiPlan.fromJson(e))
          .toList(),
    );
  }
}
